model/event/event_registry.go

```diff
@@ -68,9 +76,54 @@ func (r *eventRegistry) IsSubscribable(resourceType, eventType string) bool {
 		EventType:    eventType,
 	}
 
-	allowSubs := r.isSubscribable[e]
+	return r.isSubscribable[e]
+}
+
+// RegisterExtraData allows a pair of (resourceType, trigger) to deserialize
+// data from the ExtraData field. Once registered, the data is required.
+func (r *eventRegistry) RegisterExtraData(resourceType, triggerName string, i interface{}) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if i == nil {
+		panic(fmt.Sprintf("attempted to register nil extra data for event '%s', trigger: '%s' more than once", resourceType, triggerName))
+	}
+
+	e := extraDataKey{
+		ResourceType: resourceType,
+		Trigger:      triggerName,
+	}
+
+	_, ok := r.extraData[e]
+	if ok {
+		panic(fmt.Sprintf("attempted to register extra data for event '%s', trigger: '%s' more than once", resourceType, triggerName))
+	}
+
+	if reflect.TypeOf(i).Kind() == reflect.Struct {
+		panic(fmt.Sprintf("extra data must be a pointer to a struct, saw Kind: %s", reflect.TypeOf(i).Kind().String()))
+	}
+
+	r.extraData[e] = i
+}
+
+// GetExtraData initializes extra data for the given pair (resourceType, trigger)
+// The type returned is guaranteed to be not nil
+func (r *eventRegistry) GetExtraData(resourceType, triggerName string) interface{} {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
 
-	return allowSubs
+	e := extraDataKey{
+		ResourceType: resourceType,
+		Trigger:      triggerName,
+	}
+
+	data, ok := r.extraData[e]
+	if !ok {
+		return nil
+	}
+
+	t := reflect.ValueOf(data).Type()
+	return reflect.New(t).Interface()
```

Comment:
why can't you return `t` here?

Reply:
That would return the same t stored in the registry, I want a new instance of the same type.